docs: improve CRD reference pages (#2426)
keptn/lifecycle-toolkit: docs/content/en/docs/yaml-crd-ref/app.md

weight: 10
---

> A `KeptnApp` is a Custom Resource Definition (CRD) that serves as a blueprint for defining and managing all the essential components, including [workloads](https://kubernetes.io/docs/concepts/workloads/) that are  necessary to operate our application within a Kubernetes cluster. This definition outlines the various pieces that make up our application, ensuring their proper orchestration and coordination.

Reviewer: ```suggestion
A `KeptnApp` resource defines and manages the essential components, including [workloads](https://kubernetes.io/docs/concepts/workloads/) that are  necessary to operate our application within a Kubernetes cluster.
This definition outlines the various pieces that make up the application,
ensuring their proper orchestration and coordination.
```
Principle of parsimony should apply, especially in reference docs.  This is a page in the "CRD Reference" so we don't need to say that each one is a CRD or spell out what a CRD is.

Line length must not exceed 120 characters and something closer to 80 characters is ideal.  This is also causing your PR to fail markdown.  Hint: you can run `make markdownlint-fix` on your local branch and it will fix a lot of these errors.  

I'm not sure that "essential components" is meaningful here at the beginning without explaining the concept of a KeptnApp.  Perhaps you should move some content from the "Usage" section up to the beginning?

I made a few suggestions but I think this needs some rewriting.  The upfront material should not be too long.  Remember, these pages are the reference material one looks at over and over again and what one mostly wants to get to is the Synopsis and Field Descriptions.  New users should start with the User Guides.

It contains information about:

- All [workloads](https://kubernetes.io/docs/concepts/workloads/) and checks
  that are associated with a Keptn application

> - A list of tasks and evaluations to be can be executed such as 

Reviewer: Extra words here...  Why do you say, "can be executed"?  If they are listed here, they will be executed.  "such as" also makes no sense -- all tasks defined in a KeptnApp are defined for either pre-deployment or post-deployment execution -- there are no other options.

> - A list of tasks and evaluations to be can be executed such as 
>   pre and post-deployment.

Reviewer 2: ```suggestion
- A list of tasks and evaluations to be can be executed 
  pre or post deployment of the workloads.
```

> - Tasks which are referenced by `KeptnApp` are defined in a

Reviewer: ```suggestion
- Tasks that are referenced by `KeptnApp` are defined in a
```
If you must add the extra text here, it should be the restrictive "that" .  But I actually think this is superfluous text.

>   [KeptnTaskDefinition](taskdefinition.md) resource which helps us for more stable deployment.

Reviewer: What do you mean by "which helps us for more stable deployment"?

> - `KeptnApp` simply looks at the name of a task stored in `metadata.name` to figure out what the task is, without needing to know all the technical details of how the task is carried out by a runner. In other words, it recognizes tasks by their names and doesn't get bogged down in the specifics of how they work.

Reviewer: This needs to be broken up into multiple source lines and I would omit "simply".  But I'm not sure the information is useful or completely accurate.

- **apiVersion** -- API version being used.
- **kind** -- Resource type.
   Must be set to `KeptnApp`.

- **metadata** -- Information about the resources.

>   - **name** -- This field is used to specify the unique name of the application or resource being defined. The name must follow the guidelines outlined in the [Kubernetes Object Names and IDs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names) specification.

Reviewer: Why do you say "the application or resource being defined" rather than sticking to the "Unique name of this application."  But good that you added the link to the k8s guidelines -- it was an oversight that it wasn't there.

>   - **namespace** -- Isolates resources in Kubernetes, aiding organization and preventing naming conflicts across clusters, enhancing multi-tenant and collaborative cluster management. 

Reviewer: This kind of defines what a namespace is but it doesn't define what information goes into this field for this CRD and what the implications are.  You might get some help from https://lifecycle.keptn.sh/docs/install/k8s/#how-many-namespaces.
